### PR TITLE
KAKFA-9503: Fix TopologyTestDriver output order

### PR DESCRIPTION
**Pull request by vvcephei**
Because of the recursive invocations of pipeRecord, records in recursive topologies
would be appended to the TopologyTestDriver output buffer in depth-first order,
rather than breadth-first, which is how KafkaStreams would do it.

This PR changes the processing algorithm from an (implicitly, via recursion) stack-based
one to a queue-based one more similar to how KafkaSteams actually executes.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by vvcephei** [comment]
Hey @mjsax , do you have a chance to take another look at this fix? I've isolated it from #8015 (the fix for KAFKA-9500).

I also cleaned it up quite a bit in response to the feedback on #8015.

@bbejeck , if you have time, I'd also appreciate your feedback. I tagged you because KAFKA-9503 blocks KAFKA-9500, which itself blocks the 2.4.1 release.

**Comment by vvcephei** [comment]
Oh man, getting some super flaky tests again:
    kafka.admin.DescribeConsumerGroupTest.testDescribeGroupMembersWithShortInitializationTimeout
kafka.admin.DescribeConsumerGroupTest.testDescribeGroupWithShortInitializationTimeout
kafka.admin.ResetConsumerGroupOffsetTest.testResetOffsetsExportImportPlan

Retest this, please.

**Comment by vvcephei** [comment]
One flaky test:
* kafka.admin.DescribeConsumerGroupTest.testDescribeGroupWithShortInitializationTimeout

I'm not going to bother to re-run it.

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (mjsax) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: Not sure why you flip this condition?

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (mjsax) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: nit: add a blank line before this one to make it easier to read

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (mjsax) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: Should we move this check out of this method to the caller? It's only called twice and one caller does this check outside already.

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (mjsax) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: Should we add the same check for `topicName`, `timestamp`, and `headers` ?

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (mjsax) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: There is already `StreamTask#numBuffered()` -- should we reuse this one instead of adding a new method?

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java
Comment (mjsax) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: Why do we not use a `List`?

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java
Comment (mjsax) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java:1561]: Why do we test the same thing twice?

**Review by mjsax — COMMENTED** [review]
Thanks for the PR. This is already simpler. It's still hard to follow the flow -- the interplay/call-structure of the new methods is quite complex -- do you think we can simplify further?

Btw: there is type in a test method `shouldPunctuateIfEvenTimeAdvances` (can you fix it on the side :) )

Also given the new structure, we might add one more test to ensure that we process "recursive" records for global topics correctly (we have `shouldForwardRecordsFromSubtopologyToSubtopology` already and in the old structure with the simple "pipeInput-loop" it was sufficient -- now the code is more complex as we have a different code path for global-records and thus it might be good to add a test for the case in which we write into a global input topic)

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (vvcephei) [reply streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: Just to improve readability.

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (vvcephei) [reply streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: I can; I only added this check because it was never possible to go into this logic before with a null topicPartition (because it was inside a condition block that `topicPartition != null`), but now that it's a separate method, future (buggy) code changes could result in passing a `null` here. We'd just get an NPE later on, but I wanted to check it up front on a line of its own to decrease debugging time.

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (vvcephei) [reply streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: That one is private. I considered it, but it seemed silly to expose `integer numBuffered()` when the only thing we'd do with it is `while (numBuffered() > 0)`. If we have another reason to expose the integer in the future, we can merge the use cases later on.

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java
Comment (vvcephei) [reply streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: I also verified the list later on (as I see you noticed ;) ).

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java
Comment (vvcephei) [reply streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java:1561]: It's not exactly the same.

We do want the "final" output to be the correct end-state of the operation, which the map test verifies very succinctly. Since this is a table operation, verifying the result as a map is most appropriate.

But we also don't want anything extra in the output topic, which the second test verifies.

It's true that the map test is redundant with the last element in the list test; it's just telling the narrative of what we expect.

Comment (vvcephei) [reply streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java:1561]: I've added comments to make this narrative more apparent.

**Comment by vvcephei** [comment]
Hey @mjsax , thanks for the review! I've addressed your comments.

Regarding your high-level feedback about whether the workflow could be simpler, I'm not sure. I'm probably blinded by the same mentality that led me to write this code to begin with, but it seems pretty straightforward to me. I.e., Kafka Streams is a queue-based processing system, so the algorithm of "enqueue inputs, then work down the queue (which might add more work to the end of the queue) until it's empty" seems like a pretty natural approach. And it's also exactly the same algorithm that KafkaStreams would execute.

If you have other ideas, please do share; I'm just coming up empty in response to the question of simplifying it further.

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (mjsax) [reply streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: Personally I think it reduced readability 😛

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java
Comment (mjsax) [reply streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: But the list-test is a superset of the map test? Seems to be redundant to me.

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java
Comment (mjsax) [reply streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java:1561]: > Since this is a table operation

The test uses the PAPI -- those semantic cannot be inferred -- I am fine with leaving the redundant code if you like it better that way -- but it does not improve test coverage and is confusing IMHO.

Comment (vvcephei) [reply streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java:1561]: Ah! I was still thinking about the foreign-key join test. You're right.

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (mjsax) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: `ProducerRecord` is actually guarantee to have a non-null timestamp -- not sure if we need this check -- actually similar for other fields like topic or header

**Comment by mjsax** [comment]
I think we could simplify the logic by "encapsulated" the logic of certain methods:

- `enqueueTaskRecords` is already nicely encapsulated: it only does one thing
- similar `processGlobalRecord` does only one thing
- however, the interplay of `processAllProcessableRecords()` and `captureOutputRecords()` is not clear IMHO

I would suggest that `processAllProcessableRecords()` only drains the task queue, and after it returns all output records would still be in the mocked-producer record-buffer (instead of re-enqueuing inputs already; with the side effect that output records are put into `TestOutputTopic` and global-record are processed "on the side").

Furthermore, `captureOutputRecords()` (we could rename it, maybe to `captureOutputRecordsAndContinueProcessing` ?) has the duty to do the following in a loop:
 - get record from the producer and
   - put them into the `TestOutputTopics`
   - re-enqueue them if necessary
   - for the "global task" we process immediately
- trigger processing all newly enqueued input records

The loop terminates if no output was re-enqueued.

This makes `captureOutputRecordsAndContinueProcessing` the "main driver" while all other methods do one thing. It also allows to make a single method call to `captureOutputRecordsAndContinueProcessing` within `advanceWallClockTime` and `close()` instead of two.

I illustrated my idea in this branch: https://github.com/mjsax/kafka/commit/30ef084f682798171984f5523a941827cb2e330e

It's to some extent cosmetic and I leave it up to you to adopt what I suggest or not. My commit also does some other nice side cleanup (improved naming and fixing some generics). Let me know what you think.

Feel free to merge this PR if Jenkins in green.

**Review by mjsax — APPROVED** [review]

**Comment by vvcephei** [comment]
Hi @mjsax , thanks so much for the feedback and the demonstration branch. I've copied over all your cleanups and also made some changes in light of your thoughtful feedback above.

I'm not 100% on the same page about everything, but a lot of what you said resonates with me. I think the sticking point for with me is that typically in queue-based algorithms, you just enqueue the starting work item, and then enter a processing loop (polling one unit of work, processing, maybe enqueuing resulting work) until the queue is empty.

However, I'm realizing that I've created some unfortunate dissonance by keeping the internal method names similar while changing their roles substantially. Accordingly, I've renamed the outer loop portion to `completeAllProcessableWork` and the "output collection" method to `captureOutputsAndReEnqueueInternalResults`. The high-level flow is now:
1. enqueue input
2. loop while there is work enqueued:
    1. process the head of the queue
    2. capture any outputs, and re-enqueue any outputs for internal topics 

Those two methods replace what used to be co-recursive peers, but now, the `completeAllProcessableWork` method is the "main loop" and `captureOutputsAndReEnqueueInternalResults` is just a helper.

Just comparing the structure of this algorithm to what you wrote above, it seems very close in spirit to what you were proposing.

I very much liked your suggestion to not need to call the helper method before the "main loop" method when punctuating or closing, so I've internalized it inside `completeAllProcessableWork` itself.

While reflecting on your feedback, I actually realized unrelated: if the `max.task.idle.ms` option is supplied to the TopologyTestDriver, we could end up with a non-terminating test. Since wall-clock time doesn't advance automatically, we could get stuck with work enqueued that `task.process` would never be able to process. I've added the new test `shouldRespectTaskIdling` to cover this case, and expanded the loop condition above to `while (task.hasRecordsQueued() && task.isProcessable(mockWallClockTime.milliseconds()))`. I'm pretty sure this was a bug that predates this PR, caused by an earlier refactoring I did (on trunk only) of the Task interface. That refactoring changed `task.process()` (unconditional) to `task.process(wallClockTime)` (return early if the task isn't processable).

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:447]: Exposed this again for use in TopologyTestDriver

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (vvcephei) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java:527]: Here's the loop condition I added to let the loop terminate when task idling limits our ability to process enqueued records.

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java
Comment (vvcephei) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java:1696]: Here's the new test I added to make sure TopologyTestDriver works properly with task idling.

**Review by vvcephei — COMMENTED** [review]
Just calling out the "task idling" bug I'm fixing on the side:

**Comment by guozhangwang** [comment]
retest this please

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (mjsax) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java:541]: I think we should add a INFO (or WARN) log statement for the case that there is buffered data that cannot be processed yet and inform the use that they need to advance wall-clock time manually.

We might also have an INFO log when the TTD is created if we detect that task.idle is non-zero?

How should we handle `close()`? I think we should internally advance wall-clock time to ensure we drain all output? If we leave it to the user the might not be able to drain all records if there is a loop in the dataflow (what would required to advance wall-clock time after each new enqueue?). We would also have a test for this case if we implement it that way.

**Review by mjsax — COMMENTED** [review]
Thank @vvcephei -- I like the new version very much! The control flow of the code is clean now.

Nice find about `task.idle` -- one inline comment about it.

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (vvcephei) [reply streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java:541]: Thanks, @mjsax !

Adding both log messages as INFO seems appropriate since this is a slight behavior change, which might affect some programs. I'm not sure about WARN for the "still buffered" case, since, you'd be guaranteed to get the log message when using task.idle. I.e., we'd be giving an unavoidable warning telling you that nothing at all is wrong. "Info" should be high enough visibility.

That's a good point about `close`. How concerning would it be to just have unprocessed records while closing the TTD? That would be the cleanest approach from an implementation perspective, and it seems like the test author couldn't care that much about the results, or they would have actually advanced wall-clock time themselves to get them.

**Comment by vvcephei** [comment]
Thanks again for helping work through these issues, @mjsax . Upon reflection, it seems like we can punt the question about flushing out those left-over records in `close()`, since it's probably not too common to interact with the test driver after it has been closed anyway, ever since we added support and re-wrote the examples to use try-with-resources.

It seems like we can revisit this and also take on the issue of flushing out suppressed events in follow-up work to add some kind of `flushTopology()` operation. For now, to partially address your concern, I've added a warning log if there are still enqueued records after the driver is closed, in addition to the two other INFO-level logs.

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (mjsax) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: nit: add `()` -> `advanceWallClockTime()` ?

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (mjsax) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: What does `Such occurrences` means? Might be good to be more elaborative.

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (mjsax) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: `empty partitions` -> should we say `empty topics` ?

Just to clarify that some partitions from _different_ topics are empty?

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (mjsax) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: `can't` -> `could not` ?

I would remove `even though TopologyTestDriver is shutting down` (read cleaner this way IMHO.)

**Review by mjsax — APPROVED** [review]
Thanks @vvcephei -- some minor suggestions to improve the log statements. Feel free to merge if Jenkins passed.

Should we create a Jira for the follow up work? Do you think we need to update some JavaDocs or the docs on the web page about the new behavior?

**Comment by vvcephei** [comment]
Test failure was unrelated: kafka.admin.DeleteConsumerGroupsTest.testDeleteCmdAllGroups

**Comment by vvcephei** [comment]
Cherry-picked to 2.5 with https://github.com/apache/kafka/commit/7b71cb92b539a547a99dc6dd094f475fd55a8572

Cherry-picked to 2.4 with https://github.com/apache/kafka/commit/42062a89f2a7333263c8aff3e67b0eba47b29377
* I decided to omit the `isProcessable` check for 2.4, so we wouldn't change behavior in a bugfix release, and updated the task.idle test just to make sure that TTD will still terminate even if the task.idle config is provided.

**Comment by mjsax** [comment]
I left a couple of comment on the `2.4` cherry-pick PR -- if we don't use `task.idle` config, we should remove all the log statements about it.

**Pull request merged**
